### PR TITLE
State machine improvements

### PR DESCRIPTION
**Pull request by luigi-rosso**
Improving StateMachine api based on feedback from Felix @ SimpleClub.

### CODE REVIEW

File: lib/src/state_machine_controller.dart [hunk 1]
```diff
@@ -1,17 +1,51 @@
...
+  /// The id of the input within the context of the [StateMachine] it belongs
+  /// to.
+  int get id => _input.id;
+
+  /// The name given to this input at design time in Rive.
+  String get name => _input.name;
+
+  /// Convenience method for changing the backing value of the input. Usually
+  /// it's easier to use the various value getter/setters on the derived
+  /// version of [SMIInput] like [SMIBool], [SMINumber], and [SMITrigger].
```
Comment (creativecreatorormaybenot) [inline lib/src/state_machine_controller.dart:48]: Are you referring only to `SMITrigger.fire` in this case or am I missing a member?

Comment (creativecreatorormaybenot) [reply lib/src/state_machine_controller.dart:48]: If you mean the `value` setter on `SMIInput`, I think you can link to `[value]` directly.

File: lib/src/state_machine_controller.dart [hunk 2]
```diff
@@ -1,17 +1,51 @@
...
+enum SMIType { number, boolean, trigger }
+
+/// StateMachine Instance Input. This is the abstraction of an instanced input
+/// from the [StateMachine]. Whenever a [StateMachineController] is created, the
+/// list of inputs in the corresponding [StateMachine] is wrapped into a set of
+/// [SMIInput] objects that ensure inputs are initialized to design-time values.
+/// The implementation can now change these values freely as they are decoupled
+/// from the backing [StateMachine] and can safely be re-instanced by another
+/// controller later.
+class SMIInput<T> {
```
Comment (creativecreatorormaybenot) [inline lib/src/state_machine_controller.dart:29]: I am wondering, can / should this be an abstract class? Or is this instantiated somewhere?

```suggestion
abstract class SMIInput<T> {
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,8 @@
+
+## [0.7.2] - 2021-04-12 16:57:54
+- Breaking change! StateMachineInput has been renamed to SMIInput to follow conventions in other runtimes and clearly disambiguate between core.StateMachineInput (the backing type in Rive's core system, which is not explicitly exposed to this runtime) and the input instances which should be used by controllers in the Flutter ecosystem.
```
[the remark this answers is not in the document]
Comment (creativecreatorormaybenot) [reply CHANGELOG.md:3]: After reviewing the code, I think I understand it. My understanding is that `SMI` stands for `StateMachineInput` and `SMIInput` is the abstract base class. 

Naming decision like this are always difficult (: I do not have any nice idea for avoiding the conflict. The best I can come up with is `SMInput` and `SMTriggerInput`, `SMNumberInput`, etc. respectively.

**Review by creativecreatorormaybenot — APPROVED** [review]
I really like the API changes, thanks for integrating them!

I left some comments in case you are interested. In general, I do approve :) I do like the internal `typeKey` additions as well 👍

**Comment by luigi-rosso** [comment]
Thanks fir all the feedback! I'll reply in more detail later today. SMIInput = StateMachineInstance Input. Naming this was a nightmare but fits the StateMachineInstance pattern we have in other runtimes. In Flutter it's rolled into the controller, which is effectively a StateMachineInstance.

File: CHANGELOG.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (luigi-rosso) [reply CHANGELOG.md:3]: Yeah we chose to leave the I in as it stands for Instance. This makes it clear mostly for us when we work on the under the hood parts which one is an instance (has mutable state) vs the design time created inputs. State Machine Instance inputs are the ones that you interface with at the Flutter level that are backed by core.StateMachineInput derivates.

File: lib/src/state_machine_controller.dart [hunk 4]
```diff
@@ -1,17 +1,51 @@
...
+/// enum is exposed for convenience.
+enum SMIType { number, boolean, trigger }
+
+/// StateMachine Instance Input. This is the abstraction of an instanced input
+/// from the [StateMachine]. Whenever a [StateMachineController] is created, the
+/// list of inputs in the corresponding [StateMachine] is wrapped into a set of
+/// [SMIInput] objects that ensure inputs are initialized to design-time values.
+/// The implementation can now change these values freely as they are decoupled
+/// from the backing [StateMachine] and can safely be re-instanced by another
+/// controller later.
```
[the remark this answers is not in the document]
Comment (luigi-rosso) [reply lib/src/state_machine_controller.dart:28]: Yes, it only returns a reference to an existing instanced input (created during construction of the controller). This ensures you can call it multiple times and get back the same instance.

File: lib/src/state_machine_controller.dart [hunk 5]
```diff
@@ -20,15 +54,89 @@ class StateMachineInput<T> {
...
+  void advance() => change(false);
 }
 
 /// An AnimationController which controls a StateMachine and provides access to
 /// the inputs of the StateMachine.
 class StateMachineController extends core.StateMachineController {
+  final List<SMIInput> _inputs = <SMIInput>[];
+
+  /// A list of inputs available in the StateMachine.
+  Iterable<SMIInput> get inputs => _inputs;
```
[the remark this answers is not in the document]
Comment (luigi-rosso) [reply lib/src/state_machine_controller.dart:118]: This was definitely a poor oversight in the first implementation. Thanks for the feedback on all of this!

File: CHANGELOG.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (creativecreatorormaybenot) [reply CHANGELOG.md:3]: Thanks, makes sense.

File: lib/src/state_machine_controller.dart (hunk 2 above)
Comment (luigi-rosso) [reply lib/src/state_machine_controller.dart:29]: That's a great point, it should totally be abstract.

**Pull request merged**